### PR TITLE
Reduce memory use for PULUMI_OPTIMIZED_CHECKPOINT_PATCH

### LINKED ISSUE #11653: PULUMI_OPTIMIZED_CHECKPOINT_PATCH temporarily allocates 100x the checkpoint memory
### What happened?

An attempt to roll out  PULUMI_OPTIMIZED_CHECKPOINT_PATCH as default behavior resulted in a P0 indicating that users are experiencing OOM issues with their Pulumi programs https://github.com/pulumi/pulumi/issues/11650

A careful analysis of a repro case that was submitted in private and cannot be shared here fully has pinpointed the issue to the diff algorithm. Here are some excerpts of the pprof memory profiler on a program that works on a 6MB uncompressed JSON checkpoint:

```
github.com/hexops/gotextdiff/myers.shortestEditSequence
/Users/t0yv0/go/pkg/mod/github.com/hexops/gotextdiff@v1.0.3/myers/diff.go

  Total:      3.43GB     3.43GB (flat, cum) 98.88%
    148            .          .           } 
    149            .          .            
    150            .          .           // shortestEditSequence returns the shortest edit sequence that converts a into b. 
    151            .          .           func shortestEditSequence(a, b []string) ([][]int, int) { 
    152            .          .           	M, N := len(a), len(b) 
    153       1.96MB     1.96MB           	V := make([]int, 2*(N+M)+1) 
    154            .          .           	offset := N + M 
    155       2.88MB     2.88MB           	trace := make([][]int, N+M+1) 
    156            .          .            
    157            .          .           	// Iterate through the maximum possible length of the SES (N+M). 
    158            .          .           	for d := 0; d <= N+M; d++ { 
    159       3.42GB     3.42GB           		copyV := make([]int, len(V)) 
    160            .          .           		// k lines are represented by the equation y = x - k. We move in 
    161            .          .           		// increments of 2 because end points for even d are on even k lines. 
    162            .          .           		for k := -d; k <= d; k += 2 { 
    163            .          .           			// At each point, we either go down or to the right. We go down if 
    164            .          .           			// k == -d, and we go to the right if k == d. We also prioritize 

github.com/hexops/gotextdiff/myers.operations
/Users/t0yv0/go/pkg/mod/github.com/hexops/gotextdiff@v1.0.3/myers/diff.go

  Total:           0     3.43GB (flat, cum) 98.88%
     47            .          .           func operations(a, b []string) []*operation { 
     48            .          .           	if len(a) == 0 && len(b) == 0 { 
     49            .          .           		return nil 
     50            .          .           	} 
     51            .          .            
     52            .     3.43GB           	trace, offset := shortestEditSequence(a, b) 
     53            .          .           	snakes := backtrack(trace, len(a), len(b), offset) 
     54            .          .            
     55            .          .           	M, N := len(a), len(b) 
     56            .          .            
     57            .          .           	var i int 
     58            .          .           	solution := make([]*operation, len(a)+len(b)) 
     59            .          .            
     60            .          .           	add := func(op *operation, i2, j2 int) { 
     61            .          .           		if op == nil { 
     62            .          .           			return 
     63            .          .           		} 

github.com/hexops/gotextdiff/myers.ComputeEdits
/Users/t0yv0/go/pkg/mod/github.com/hexops/gotextdiff@v1.0.3/myers/diff.go

  Total:           0     3.43GB (flat, cum) 98.95%
     15            .          .           // Sources: 
     16            .          .           // https://blog.jcoglan.com/2017/02/17/the-myers-diff-algorithm-part-3/ 
     17            .          .           // https://www.codeproject.com/Articles/42279/%2FArticles%2F42279%2FInvestigating-Myers-diff-algorithm-Part-1-of-2 
     18            .          .            
     19            .          .           func ComputeEdits(uri span.URI, before, after string) []diff.TextEdit { 
     20            .     3.43GB           	ops := operations(splitLines(before), splitLines(after)) 
     21            .          .           	edits := make([]diff.TextEdit, 0, len(ops)) 
     22            .          .           	for _, op := range ops { 
     23            .          .           		s := span.New(uri, span.NewPoint(op.I1+1, 1, 0), span.NewPoint(op.I2+1, 1, 0)) 
     24            .          .           		switch op.Kind { 
     25            .          .           		case diff.Delete: 
```

Essentially the root cause here is feeding too many newlines into the Myers algorithm. In the example the JSON document has 62641 newlines.

Note that Pulumi has code that injects these newlines for PULUMI_OPTIMIZED_CHECKPOINT_PATCH specifically to enable this Myers algorithm to perform well in detecting diffs; therefore we have control over how many newlines to generate. 

https://github.com/pulumi/pulumi/pull/11568 incidentally seems to fix this issue by reducing the newlines to one-per-resource.

A potential fix here is to extract only the parts of 11568 relevant to reducing newline counts.
 

### Steps to reproduce

This is difficult to reproduce since built-in Pulumi profiling capability performs a manual GC before dumping the memory profile, which reclaims the space. I've reproduced by writing a custom test case to stress-test the lower level functions, and collect a heap profile without manual GC when a threshold is crossed.

### Expected Behavior

Pulumi uses memory 2-3x the size of the checkpoint at most.

### Actual Behavior

Pulumi uses memory 100x the size of the checkpoint at most.

### Output of `pulumi about`

_No response_

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by t0yv0**
# Description

Fixes #11653

Inject fewer newlines, and per @blampe suggestion do not inject more than 1024 newlines even if we have more than 1024 resources. This limits Myers matrix to 1024x1024 max size which should be 8MB or so.

Promotes jsoniter from indirect to direct dependency. It's easier to work with for preserving indents and doing newline injection.

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/backend/httpstate/client/marshal.go [hunk 1]
```diff
@@ -17,19 +17,102 @@ package client
...
+
+func (c *marshalUntypedDeployment) Write(w io.Writer) error {
+	cfg := jsonIterConfig
+	stream := cfg.BorrowStream(w)
+	defer cfg.ReturnStream(stream)
+	err := c.writeToStream(stream)
+	return err
+}
+
+func (c *marshalUntypedDeployment) writeToStream(stream *jsoniter.Stream) error {
```
Comment (RobbieMcKinstry) [inline pkg/backend/httpstate/client/marshal.go:64]: I have no comprehension of the incantation order here 😅 For example, what does `WriteMore()` do? Why is it necessary? Maybe a single line comment somewhere at the top of the method would help? Not a big deal either way, I feel like someone coming across this is going to have to read the API docs anyway.

File: pkg/backend/httpstate/client/marshal.go [hunk 2]
```diff
@@ -17,19 +17,102 @@ package client
...
+		stream.WriteMore()
+		stream.WriteObjectField("secrets_providers")
+		stream.WriteVal(deployment.SecretsProviders)
+	}
+	if err = stream.Flush(); err != nil {
+		return err
+	}
+	nResources := len(deployment.Resources)
+
+	maxNewlines := 1024 - 2
```
Comment (RobbieMcKinstry) [inline pkg/backend/httpstate/client/marshal.go:90]: Pull this into a file-level constant?

File: pkg/backend/httpstate/client/marshal.go [hunk 3]
```diff
@@ -17,19 +17,102 @@ package client
...
+	newlinePeriod := int(math.Ceil(float64(nResources) / float64(maxNewlines)))
```
Comment (RobbieMcKinstry) [inline pkg/backend/httpstate/client/marshal.go:91]: Oh, same here I suppose.

File: pkg/backend/httpstate/snapshot.go [hunk 4]
```diff
@@ -102,23 +101,10 @@ func (persister *cloudSnapshotPersister) saveDiff(ctx context.Context,
 }
 
 func (persister *cloudSnapshotPersister) saveFullVerbatim(ctx context.Context,
-	differ *deploymentDiffState, deployment *apitype.UntypedDeployment, token string) error {
```
Comment (RobbieMcKinstry) [inline pkg/backend/httpstate/snapshot.go:105]: Nice! This really cleans up the function by encapsulating the logic in `marshalUntypedDeployment`

File: pkg/backend/httpstate/snapshot_test.go [hunk 5]
```diff
@@ -207,17 +229,8 @@ func TestCloudSnapshotPersisterUseOfDiffProtocol(t *testing.T) {
...
-		"\"column\":1,\"offset\":-1}},\"NewText\":\"\"},{\"Span\":{\"uri\":\"\",\"start\":"+
-		"{\"line\":14,\"column\":1,\"offset\":-1},\"end\":{\"line\":15,\"column\":1,"+
-		"\"offset\":-1}},\"NewText\":\"\"},{\"Span\":{\"uri\":\"\",\"start\":{\"line\":15,"+
-		"\"column\":1,\"offset\":-1},\"end\":{\"line\":16,\"column\":1,\"offset\":-1}},"+
-		"\"NewText\":\"\"},{\"Span\":{\"uri\":\"\",\"start\":{\"line\":16,\"column\":1,"+
-		"\"offset\":-1},\"end\":{\"line\":17,\"column\":1,\"offset\":-1}},\"NewText\":\"\"}]",
-		string(req3.DeploymentDelta))
-	assert.Equal(t, "5f2fd84a225e7c1895528b4b9394607d6b39aefa4ee74f57e018dadcb16bf2e2", req3.CheckpointHash)
+	assertEqual("req3", req3.DeploymentDelta)
+	assertEquals("req3.hash", req3.CheckpointHash)
```
Comment (RobbieMcKinstry) [inline pkg/backend/httpstate/snapshot_test.go:233]: God bless you for this <3

**Review by RobbieMcKinstry — APPROVED** [review]
Nice! Looks good to me! Sorry I'm a menace with pedantic nits :( They're non-blocking though.

File: pkg/backend/httpstate/client/marshal.go (hunk 1 above)
Comment (t0yv0) [reply pkg/backend/httpstate/client/marshal.go:64]: WriteMore writes a "," .. this is a pretty low-level API.

File: pkg/backend/httpstate/client/marshal.go [hunk 6]
```diff
@@ -17,19 +17,102 @@ package client
...
 )
 
-// Unlike json.Marshal preserves possible indentation in req.Deployment.
-func marshalVerbatimCheckpointRequest(req apitype.PatchUpdateVerbatimCheckpointRequest) (json.RawMessage, error) {
-	sentinel := []byte(`"*"`)
-	cp := req
-	cp.UntypedDeployment = sentinel
-	pattern, err := json.Marshal(cp)
-	if err != nil {
+var jsonIterConfig = jsoniter.Config{SortMapKeys: true}.Froze()
```
[the remark this answers is not in the document]
Comment (t0yv0) [reply pkg/backend/httpstate/client/marshal.go:28]: It's a strange lib. Froze() allocates state/cache/pools. This needs to be a module-level var I think for best results (reuse of the pools).

**Comment by t0yv0** [comment]
bors merge

File: pkg/backend/httpstate/client/marshal.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (RobbieMcKinstry) [reply pkg/backend/httpstate/client/marshal.go:30]: SGTM! IMO module level variables is typically an anti-pattern, except when it isn't. Which is Golang in a nutshell... :\

This is a good example of when it isn't.

File: pkg/backend/httpstate/client/marshal.go (hunk 1 above)
Comment (RobbieMcKinstry) [reply pkg/backend/httpstate/client/marshal.go:66]: 😆

**Pull request merged**